Get rid of locking at node {publish,unpublish} operations

humblec opened the pull request:
Considering kubelet make sure node operations like unpublish/nodeunpublish
..etc can **not** be called at same time, we dont require extra locking
in nodePublish and nodeUnpublish operation from driver side. Till this commit, we were
operating at NodePublish/NodeUnpublish with a lock and it is now removed.

**RBD**:

- [x] Lock removed from nodePublish
- [x] Lock removed from nodeUnpublish


**CephFS**

- [x] Lock removed from nodePublish
- [x] Lock removed from nodeUnpublish


Fixes: #https://github.com/ceph/ceph-csi/issues/2123
Updates: https://github.com/ceph/ceph-csi/issues/2197

Thanks to @mape90 for reporting and analyzing this issue

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@Madhu-1 , fyi:
```
Step 1/6 : FROM registry.fedoraproject.org/fedora:latest
Get https://registry.fedoraproject.org/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
make: *** [Makefile:207: .test-container-id] Error 1
Error: Process completed with exit code 2.
```

humblec commented:
/retest all

Yuggupta27 commented:
> /retest all

Currently CI is down, so centos-ci tests will not get triggered :(

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -254,16 +254,11 @@ func (ns *NodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
+	// notice: we have been operating here with a lock while unmounting the volume
+	// which slowdown the unmount operation when there are more pods and delete performed
+	// at almost same time. Considering we have enough safe-guards on CO side, we can
+	// operate outside the critical section.
```
This comment is not needed, and  probably more confusing than it can be. No need to mention issues (slowdown), just explain *why* it is safe to not use locking here.

It would be better to explain the working of NodeUnpublishVolume in the comment above the function, and maybe mention that locking is not required (why?) for unmounting.

humblec replied:
sure... let me address this in a detail or change the comment with the sequence ..etc

nixpanic reviewed (commented):
Nits:

- "Getrid" in the subject of the PR is not a word
- `Fix: #...` is not a GitHub annotation in commit messages, use `Fixes: #...`
- Both commit messages should have a description, not only one. You can use the same description for both if you like.

Madhu-1 commented:
@humblec as we are depending on the CO for safeguarding then we don't need locking at the nodepublish also?

nixpanic commented:
On Thu, Jun 10, 2021 at 12:24:50AM -0700, Madhu Rajanna wrote:
> @humblec as we are depending on the CO for safeguarding then we don't need locking at the nodepublish also?

Agreed! NodeStageVolume should only need this for non-RWX volumes?

humblec commented:
/retest all

humblec commented:
/retest all

Madhu-1 commented:
> On Thu, Jun 10, 2021 at 12:24:50AM -0700, Madhu Rajanna wrote: @humblec as we are depending on the CO for safeguarding then we don't need locking at the nodepublish also?
> Agreed! NodeStageVolume should only need this for non-RWX volumes?

Yes correct, If CO takes care of it then we don't need it in NodeStage and Unstage also.

humblec commented:
@nixpanic @Madhu-1 there could be various locks we can get rid of or adjust. The #2197 list out the operations. I would like to keep this one at a time if possible, especially when there is a mix of 'global' and 'local' operations like node*Publish /Node*stage. It also help to seggregate or take an action properly.  NodeUnpublish case is handled here, and followup PRs will be filed accordingly. PTAL.. Thanks !

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

yati1998 reviewed (approved):
LGTM

Madhu-1 commented:
IMO the issue in this area cannot be caught easily. @humblec with how many pods and volumes this has been tested and did you see any warning/error logs in the kubelet after scaling up/down application pods several times?

humblec commented:
The general test we have with our `e2e` test kind of prove that, there is no `lockup` we come across at least on current tests. Additionally, I have been trying to make use of our scale test CI to run it on top of this PR, however not much luck, the main reason for that is, the CI test suite we have is tightly coupled to `ODF` or `downstream`. Planning to try some more to make it work so that we are sure, there is no 'regression' and also we can do some benchmark on the performance improvement we get with this change.  I will update here how it goes.

humblec commented:
[Update] The test results are good at my end. However trying to make use of scale testing by taking help from Perf team to really validate the improvement this brings us. Will update here how it goes.

humblec commented:
FYI, The global mount operations locking has been taken out and the test image is `quay.io/humble/cephcsi:publish` which has been undergoing the testing for a while now.

humblec commented:
@ekuric would you like to update  your findings as you are helping us with scale testing to verify the improvement with this fix image?

ekuric commented:
> @ekuric would you like to update your findings as you are helping us with scale testing to verify the improvement with this fix image?

I tested images you provided and I can see below observations

- Create time: It takes approximately 420s ( ~7m ) to start 1000 pods with PVC attached ( from cephfs or ceph rbd storage class ).
-  Delete time: Pods created in previous steps were in project - eg. projectX. For delete tests we ran `oc delete project projectX ` and measured time how long it takes before all devices disappear from storage backend  ( for ceph rbd `rbd ls storagecluster-cephblockpool | wc -l` and for cephfs `ceph fs subvolume ls storagecluster-cephfilesystem csi |grep name | wc -l` ). In both these cases it took between 280 - 300 seconds before all volumes are cleaned on storage side.
- During tests , there were not noticed issues with ceph cluster itself , it was operating without  ( noticeable )  issues. We can for purpose of next release run more extensive tests if necessary.

humblec commented:
> > @ekuric would you like to update your findings as you are helping us with scale testing to verify the improvement with this fix image?
> 
> I tested images you provided and I can see below observations
> 
>     * Create time: It takes approximately 420s ( ~7m ) to start 1000 pods with PVC attached ( from cephfs or ceph rbd storage class ).
> 
>     * Delete time: Pods created in previous steps were in project - eg. projectX. For delete tests we ran `oc delete project projectX ` and measured time how long it takes before all devices disappear from storage backend  ( for ceph rbd `rbd ls storagecluster-cephblockpool | wc -l` and for cephfs `ceph fs subvolume ls storagecluster-cephfilesystem csi |grep name | wc -l` ). In both these cases it took between 280 - 300 seconds before all volumes are cleaned on storage side.
> 
>     * During tests , there were not noticed issues with ceph cluster itself , it was operating without  ( noticeable )  issues. We can for purpose of next release run more extensive tests if necessary.

This is awesome!!!  :tada: Thanks @ekuric for doing this, much appreciated.  

In short, `1000 pods` delete time has been improved from `980seconds` to `~280` seconds ! Thats huge improvement.  :+1:

humblec commented:
/retest  ci/centos/mini-e2e/k8s-1.19

humblec commented:
/retest all

humblec commented:
@Madhu-1 @nixpanic considering we have a good improvement here with some test results, can we get this in ?

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -145,11 +145,8 @@ func (ns *NodeServer) NodeStageVolume(
-		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volID)
-	}
-	defer ns.VolumeLocks.Release(volID)
+	// considering kubelet make sure node operations like publish/stage...etc can not be called
```
@humblec are we sure that kubelet is taking some central locking when it comes to RWO on multiple nodes. I past we have seen that RWO request comes up on two nodes parallel and if this happens we end up corrupting the data?

Madhu-1 replied:
Discard this one. I was confused with watchers

Madhu-1 reviewed (commented):
Nothing can be caught in review for this kind of PR, Hope everything is tested intensively and everything is verified from kubernetes related to stale mounts, directories, data corruption, etc, and also the node lost scenarios is also tested.

pkalever commented on internal/rbd/nodeserver.go:
```diff
@@ -145,11 +145,8 @@ func (ns *NodeServer) NodeStageVolume(
 	}
 	defer cr.DeleteCredentials()
 
-	if acquired := ns.VolumeLocks.TryAcquire(volID); !acquired {
```
As its already discussed, I'm sure we all are aware of this, but just to remind again:

Locking is needed by Volume Healer, so at least for NodeStageVolume() we cannot remove it.
See: https://github.com/ceph/ceph-csi/pull/2108

humblec replied:
yeah, we are aware. One thing which I can do here is bringing back the nodestage and nodeunstage locks back.. let me refresh the PR with that.

humblec committed:
rbd: Get rid of locking at nodeUnpublish call

Considering kubelet make sure the unstage and unpublish operations
are serialized, we dont need any extra locking in nodeUnpublish

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
cephfs: Get rid of locking at nodeUnpublish call

Considering kubelet make sure the unstage and unpublish operations
are serialized, we dont need any extra locking in nodeUnpublish

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: Get rid of locking at nodePublish

Considering kubelet make sure the stage and publish operations
are serialized, we dont need any extra locking in nodePublish

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
cephfs: Get rid of locking at nodepublish

Considering kubelet make sure the stage and publish operations
are serialized, we dont need any extra locking in nodePublish

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> Nothing can be caught in review for this kind of PR, Hope everything is tested intensively and everything is verified from kubernetes related to stale mounts, directories, data corruption, etc, and also the node lost scenarios is also tested.

@Madhu-1  @nixpanic  Sure, btw one change I have done with last refresh is that, brought back the locking in 'global' operations which could help rbd nbd driver:

The delete performance (https://github.com/ceph/ceph-csi/pull/2149#issuecomment-872922124) has been gained even with this version of the patch. so we are good. 

Multilple PVC creation/deletion , App creation, deletion in batches ....etc are performed with this version of the patch and it looks good too, apart from all the e2e tests which has been passed. 

So, requesting to get this in at this earliest , so that more testing can go through with different test suites ( downstream) too.  Thanks :+1:

Madhu-1 reviewed (dismissed):
one thing to consider here is also the stale mount/unmount commands for the same volume when we run the commands in parallel. I hope that is also tested and for all the cases we have not seen any issues on the kubelet logs.

humblec commented:
/retest  ci/centos/upgrade-tests-cephfs

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.19

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/upgrade-tests-cephfs

humblec commented:
``` 
Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

humblec commented:
/retest ci/centos/upgrade-tests-cephfs

humblec commented:
/retest ci/centos/upgrade-tests-cephfs

nixpanic reviewed (dismissed):
@mergifyio rebase

pkalever reviewed (approved):
See if there is a need to edit the commit and PR messages. Just remove references about NodeStage and NodeUnstage.

"we were operating at NodePublish/NodeUnpublish/NodeUnstage and NodeStage with a lock and it is now removed."

Rest all looks good to me.

humblec commented:
/retest ci/centos/upgrade-tests-cephfs

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.20

nixpanic commented:
> /retest ci/centos/mini-e2e/k8s-1.20

Failed with #2264 ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/1659/pipeline)).

Madhu-1 reviewed (approved):
As discussed with @humblec we don't have any problem even if we remove the lock and the mount hangs(or takes more times) and kubelet sends `n` number of requests due to timeout.

Madhu-1 commented:
@mergifyio rebase

Yuggupta27 commented:
/retest ci/centos/mini-e2e/k8s-1.21
